File: metal-4-basics-master 2/Metal4Basics/Sources/Engine/ImageBasedLight.swift
//
// Copyright 2025 Warren Moore
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import ImageIO
import Metal
import MetalKit

class ImageBasedLight {
    let diffuseCubeTexture: MTLTexture
    let specularCubeTexture: MTLTexture
    let specularMipLevelCount: Int
    let scaleAndBiasLookupTexture: MTLTexture
    let readyEvent: (any MTLEvent)

    var rotation: simd_float3x3 = matrix_identity_float3x3
    var intensity: Float = 1.0

    class func generateImageBasedLight(url fileURL: URL) async throws -> ImageBasedLight {
        let loadTask = Task.detached {
            return try await ImageBasedLightGenerator.defaultGenerator.makeLight(url: fileURL)
        }
        switch await loadTask.result {
        case .success(let light):
            return light
        case .failure(let error):
            throw error
        }
    }

    init(diffuseCubeTexture: MTLTexture,
         specularCubeTexture: MTLTexture,
         specularMipLevelCount: Int,
         scaleAndBiasLookupTexture: MTLTexture,
         readyEvent: any MTLEvent)
    {
        self.diffuseCubeTexture = diffuseCubeTexture
        self.specularCubeTexture = specularCubeTexture
        self.specularMipLevelCount = specularMipLevelCount
        self.scaleAndBiasLookupTexture = scaleAndBiasLookupTexture
        self.readyEvent = readyEvent
    }
}

class ImageBasedLightGenerator {
    enum Error: Swift.Error {
        case libraryNotFound
        case functionNotFound
        case allocationFailed
    }

    let device: MTLDevice
    let commandQueue: MTLCommandQueue
    let equirectToCubePipelineState: MTLComputePipelineState!
    let lookupTablePipelineState: MTLComputePipelineState!
    let prefilteringPipelineState: MTLComputePipelineState!

    static var defaultGenerator: ImageBasedLightGenerator = {
        guard let device = MTLCreateSystemDefaultDevice() else {
            fatalError("Metal is not supported on this device")
        }
        let commandQueue = device.makeCommandQueue()!
        do {
            return try ImageBasedLightGenerator(device: device, commandQueue: commandQueue)
        } catch {
            fatalError("Failed to create image-based light generator: \(error)")
        }
    }()

    init(device: MTLDevice, commandQueue: MTLCommandQueue) throws {
        self.device = device
        self.commandQueue = commandQueue

        guard let defaultLibrary = device.makeDefaultLibrary() else { throw Error.libraryNotFound }

        guard let equirectToCubeFunction = defaultLibrary.makeFunction(name: "CubeFromEquirectangular") else { throw Error.functionNotFound }
        equirectToCubePipelineState = try device.makeComputePipelineState(function: equirectToCubeFunction)

        guard let lutIntegration = defaultLibrary.makeFunction(name: "F0OffsetAndBiasLUT") else { throw Error.functionNotFound }
        lookupTablePipelineState = try device.makeComputePipelineState(function: lutIntegration)

        guard let prefilterFunction = defaultLibrary.makeFunction(name: "PrefilterEnvironmentMap") else { throw Error.functionNotFound }
        prefilteringPipelineState = try device.makeComputePipelineState(function: prefilterFunction)
    }

    func makeLight(url: URL) throws -> ImageBasedLight {
        let textureLoader = MTKTextureLoader(device: device)
        let hdrOptions: [MTKTextureLoader.Option : Any] = [
            .SRGB : NSNumber(value: false),
            .allocateMipmaps : NSNumber(value: false)
        ]
        let equirectTexture = try textureLoader.newTexture(URL: url, options: hdrOptions)

        let workingPixelFormat = MTLPixelFormat.rgba16Float
        let sourceHeight = equirectTexture.height
        let sourceCubeSize = min(512, sourceHeight / 2)
        let specularCubeSize = sourceCubeSize
        let diffuseCubeSize = 32
        let lookupTableSize = 512
        let specularSampleCount = 1024
        let diffuseSampleCount = 2048
        let lutSampleCount = 512

        let sourceCubeDescriptor = MTLTextureDescriptor.textureCubeDescriptor(pixelFormat: workingPixelFormat,
                                                                              size:specularCubeSize,
                                                                              mipmapped:true)
        sourceCubeDescriptor.usage = [.shaderRead, .shaderWrite]

        guard let sourceCubeTexture = device.makeTexture(descriptor: sourceCubeDescriptor) else { throw Error.allocationFailed }
        sourceCubeTexture.label = "Environment Map (Cube)"

        let cubePixelFormat = MTLPixelFormat.rgba16Float

        let specularCubeDescriptor = MTLTextureDescriptor.textureCubeDescriptor(pixelFormat:cubePixelFormat,
                                                                                size:specularCubeSize,
                                                                                mipmapped:true)
        specularCubeDescriptor.usage = [.shaderRead, .shaderWrite]
        specularCubeDescriptor.storageMode = .private

        guard let specularCubeTexture = device.makeTexture(descriptor:specularCubeDescriptor) else { throw Error.allocationFailed }
        specularCubeTexture.label = "Prefiltered Environment (GGX)"

        let diffuseCubeDescriptor = MTLTextureDescriptor.textureCubeDescriptor(pixelFormat:cubePixelFormat,
                                                                               size:diffuseCubeSize,
                                                                               mipmapped:false)
        diffuseCubeDescriptor.usage = [.shaderRead, .shaderWrite]
        diffuseCubeDescriptor.storageMode = .private

        guard let diffuseCubeTexture = device.makeTexture(descriptor:diffuseCubeDescriptor) else { throw Error.allocationFailed }
        diffuseCubeTexture.label = "Prefiltered Environment (Lambertian)"

        let lookupTextureDescriptor = MTLTextureDescriptor.texture2DDescriptor(pixelFormat: .rgba16Float,
                                                                               width:lookupTableSize,
                                                                               height:lookupTableSize,
                                                                               mipmapped:false)
        lookupTextureDescriptor.usage = [.shaderRead, .shaderWrite]
        lookupTextureDescriptor.storageMode = .private

        guard let lookupTexture = device.makeTexture(descriptor:lookupTextureDescriptor) else { throw Error.allocationFailed }
        lookupTexture.label = "DFG Lookup Table (GGX)"

        var commandBuffer = commandQueue.makeCommandBuffer()!

        var computeCommandEncoder = commandBuffer.makeComputeCommandEncoder()!

        computeCommandEncoder.setComputePipelineState(self.equirectToCubePipelineState)
        computeCommandEncoder.setTexture(equirectTexture, index:1)
        computeCommandEncoder.setTexture(sourceCubeTexture, index:0)
        computeCommandEncoder.dispatchThreads(MTLSizeMake(sourceCubeSize, sourceCubeSize, 6),
                                              threadsPerThreadgroup:MTLSizeMake(32, 32, 1))

        computeCommandEncoder.endEncoding()

        commandBuffer.commit()

        commandBuffer = commandQueue.makeCommandBuffer()!

        let mipmapCommandEncoder = commandBuffer.makeBlitCommandEncoder()!
        mipmapCommandEncoder.generateMipmaps(for: sourceCubeTexture)
        mipmapCommandEncoder.endEncoding()

        computeCommandEncoder = commandBuffer.makeComputeCommandEncoder()!

        struct PrefilteringParams {
            var distribution: UInt32
            var sampleCount: UInt32
            var roughness: Float
            var lodBias: Float
            var cubemapSize: Float
        }

        let mipLevelCount = 5
        var levelSize = specularCubeSize
        for mipLevel in 0..<mipLevelCount {
            let levelView = specularCubeTexture.__newTextureView(with: specularCubeTexture.pixelFormat,
                                                                textureType: .typeCube,
                                                                levels: NSMakeRange(mipLevel, 1),
                                                                slices: NSMakeRange(0, 6))
            var params = PrefilteringParams(distribution: 1,
                                            sampleCount: UInt32(specularSampleCount),
                                            roughness: Float(mipLevel) / Float(mipLevelCount - 1),
                                            lodBias: 0.0,
                                            cubemapSize: Float(sourceCubeSize))
            computeCommandEncoder.setComputePipelineState(prefilteringPipelineState)
            computeCommandEncoder.setBytes(&params, length:MemoryLayout.size(ofValue:params), index: 0)
            computeCommandEncoder.setTexture(sourceCubeTexture, index: 0)
            computeCommandEncoder.setTexture(levelView, index: 1)
            computeCommandEncoder.dispatchThreads(MTLSizeMake(levelSize, levelSize, 6),
                                                  threadsPerThreadgroup:MTLSizeMake(32, 32, 1))
            levelSize >>= 1
        }

        var params = PrefilteringParams(distribution: 0,
                                        sampleCount: UInt32(diffuseSampleCount),
                                        roughness: 0.0,
                                        lodBias: 0.0,
                                        cubemapSize: Float(sourceCubeSize))
        computeCommandEncoder.setComputePipelineState(prefilteringPipelineState)
        computeCommandEncoder.setBytes(&params, length:MemoryLayout.size(ofValue: params), index: 0)
        computeCommandEncoder.setTexture(sourceCubeTexture, index: 0)
        computeCommandEncoder.setTexture(diffuseCubeTexture, index: 1)
        computeCommandEncoder.dispatchThreads(MTLSizeMake(diffuseCubeSize, diffuseCubeSize, 6),
                                              threadsPerThreadgroup:MTLSizeMake(32, 32, 1))

        var sampleCount = UInt32(lutSampleCount)
        computeCommandEncoder.setComputePipelineState(lookupTablePipelineState)
        computeCommandEncoder.setBytes(&sampleCount, length:MemoryLayout.size(ofValue: sampleCount), index: 0)
        computeCommandEncoder.setTexture(lookupTexture, index: 0)
        computeCommandEncoder.dispatchThreads(MTLSizeMake(lookupTableSize, lookupTableSize, 1),
                                              threadsPerThreadgroup:MTLSizeMake(32, 32, 1))

        computeCommandEncoder.endEncoding()

        let readyEvent = device.makeEvent()!
        commandBuffer.encodeSignalEvent(readyEvent, value: 1)

        commandBuffer.commit()

        let light = ImageBasedLight(diffuseCubeTexture:diffuseCubeTexture,
                                    specularCubeTexture:specularCubeTexture,
                                    specularMipLevelCount: mipLevelCount,
                                    scaleAndBiasLookupTexture:lookupTexture,
                                    readyEvent: readyEvent)
        return light
    }
}

File: Paloma Engine/Sources/Engine/ImageBasedLight.cpp
//
//  ImageBasedLight.cpp
//  Paloma Engine
//
//  Created by Artem on 19.01.2026.
//

#include "ImageBasedLight.hpp"

ImageBasedLight::ImageBasedLight(NS::SharedPtr<MTL::Texture> diffuseCubeTexture,
                                 NS::SharedPtr<MTL::Texture> specularCubeTexture, int specularMipLevelCount,
                                 NS::SharedPtr<MTL::Texture> scaleAndBiasLookupTexture,
                                 NS::SharedPtr<MTL::Event> readyEvent) : diffuseCubeTexture(diffuseCubeTexture), specularCubeTexture(specularCubeTexture), specularMipLevelCount(specularMipLevelCount),scaleAndBiasLookupTexture(scaleAndBiasLookupTexture), readyEvent(readyEvent)
{}

void ImageBasedLight::generateImageBasedLight(const std::string &url, MTL::Device *pDevice, std::function<void (ImageBasedLight *, NS::Error *)> completion)
{
//    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
//        
//        
//        ImageBasedLightGenerator* pGenerator = ImageBasedLightGenerator::Default(pDevice);
//        
//        ImageBasedLight* pLight = pGenerator->makeLight(url);
//        
//        if (pLight) {
//            completion(pLight, nullptr);
//        } else {
//            completion(nullptr, nullptr);
//        }
//    });
//    
    ImageBasedLightGenerator* pGenerator = ImageBasedLightGenerator::Default(pDevice);
      
      // makeLight делает тяжелую работу, но это безопасно для Main Thread при инициализации
      ImageBasedLight* pLight = pGenerator->makeLight(url);
      
      if (pLight) {
          // Вызываем callback сразу
          completion(pLight, nullptr);
      } else {
          NS::Error* pError = NS::Error::error(NS::String::string("Failed to load IBL", NS::UTF8StringEncoding), 0, nullptr);
          completion(nullptr, pError);
      }
}

ImageBasedLightGenerator* ImageBasedLightGenerator::Default(MTL::Device* pDevice) {
    static ImageBasedLightGenerator* s_pGenerator = nullptr;
    static std::once_flag s_once;
    
    std::call_once(s_once, [pDevice](){
        // Создаем отдельную очередь для генерации, чтобы не тормозить кадры
        MTL::CommandQueue* pQueue = pDevice->newCommandQueue();
        s_pGenerator = new ImageBasedLightGenerator(pDevice, pQueue);
        pQueue->release();
    });
    
    return s_pGenerator;
}

ImageBasedLightGenerator::ImageBasedLightGenerator(MTL::Device* pDevice, MTL::CommandQueue* pCommandQueue)
: _pDevice(NS::RetainPtr(pDevice))
, _pCommandQueue(NS::RetainPtr(pCommandQueue))
{
    auto pLibrary = NS::TransferPtr(_pDevice->newDefaultLibrary());
    
    auto loadPSO = [&](const char* funcName) -> NS::SharedPtr<MTL::ComputePipelineState> {
        auto pFuncStr = NS::String::string(funcName, NS::UTF8StringEncoding);
        auto pFunc = NS::TransferPtr(pLibrary->newFunction(pFuncStr));
        NS::Error* pError = nullptr;
        return NS::TransferPtr(_pDevice->newComputePipelineState(pFunc.get(), &pError));
    };
    _pEquirectToCubePSO = loadPSO("CubeFromEquirectangular");
    _pLookupTablePSO    = loadPSO("F0OffsetAndBiasLUT");
    _pPrefilteringPSO   = loadPSO("PrefilterEnvironmentMap");
}

ImageBasedLight* ImageBasedLightGenerator::makeLight(const std::string &path) {
    using namespace MTL;
    // 1. Loading Base Texture
    auto textureLoader = NS::TransferPtr(MTK::TextureLoader::alloc()->init(_pDevice.get()));
    
    auto srgbKey = MTK::TextureLoaderOptionSRGB;
    auto mipKey = MTK::TextureLoaderOptionAllocateMipmaps;
    auto valFalse = NS::Number::number(false);
    NS::Object* keys[] = { (NS::Object*)srgbKey, (NS::Object*)mipKey };
    NS::Object* values[] = { (NS::Object*)valFalse, (NS::Object*)valFalse };
    
    // FIX: dictionary returns autoreleased object -> No TransferPtr
    NS::Dictionary* hdrOptions = NS::Dictionary::dictionary(values, keys, 2);
    
    NS::Error* error = nullptr;
    // FIX: fileURLWithPath returns autoreleased -> No TransferPtr
    NS::URL* url = NS::URL::fileURLWithPath(NS::String::string(path.c_str(), NS::UTF8StringEncoding));
    
    auto equirectTexture = NS::TransferPtr(textureLoader->newTexture(url, hdrOptions, &error));
    
    if (!equirectTexture) {
        printf("Error loading HDR: %s\n", error->localizedDescription()->utf8String());
        return nullptr;
    }
    
    // Параметры
    PixelFormat workingPixelFormat = PixelFormatRGBA16Float;
    NS::UInteger sourceHeight = equirectTexture->height();
    NS::UInteger sourceCubeSize = std::min((NS::UInteger)512, sourceHeight / 2);
    NS::UInteger specularCubeSize = sourceCubeSize;
    NS::UInteger diffuseCubeSize = 32;
    NS::UInteger lookupTableSize = 512;
    NS::UInteger specularSampleCount = 1024;
    NS::UInteger diffuseSampleCount = 2048;
    NS::UInteger lutSampleCount = 512;
    
    // 2. Texture Descriptors & Allocation
    // FIX: factory methods return autoreleased -> No TransferPtr
    auto sourceCubeDescriptor = TextureDescriptor::textureCubeDescriptor(workingPixelFormat, sourceCubeSize, true);
    sourceCubeDescriptor->setUsage(TextureUsageShaderRead | TextureUsageShaderWrite);
    auto sourceCubeTexture = NS::TransferPtr(_pDevice->newTexture(sourceCubeDescriptor));
    sourceCubeTexture->setLabel(NS::String::string("Environment Map (Cube)", NS::UTF8StringEncoding));
    
    auto specularCubeDescriptor = TextureDescriptor::textureCubeDescriptor(workingPixelFormat, specularCubeSize, true);
    specularCubeDescriptor->setUsage(TextureUsageShaderRead | TextureUsageShaderWrite);
    specularCubeDescriptor->setStorageMode(StorageModePrivate);
    auto specularCubeTexture = NS::TransferPtr(_pDevice->newTexture(specularCubeDescriptor));
    specularCubeTexture->setLabel(NS::String::string("Prefiltered Environment (GGX)", NS::UTF8StringEncoding));
    
    auto diffuseCubeDescriptor = TextureDescriptor::textureCubeDescriptor(workingPixelFormat, diffuseCubeSize, false);
    diffuseCubeDescriptor->setUsage(TextureUsageShaderRead | TextureUsageShaderWrite);
    diffuseCubeDescriptor->setStorageMode(StorageModePrivate);
    auto diffuseCubeTexture = NS::TransferPtr(_pDevice->newTexture(diffuseCubeDescriptor));
    diffuseCubeTexture->setLabel(NS::String::string("Prefiltered Environment (Lambertian)", NS::UTF8StringEncoding));
    
    auto lookupTextureDescriptor = TextureDescriptor::texture2DDescriptor(PixelFormatRGBA16Float, lookupTableSize, lookupTableSize, false);
    lookupTextureDescriptor->setUsage(TextureUsageShaderRead | TextureUsageShaderWrite);
    lookupTextureDescriptor->setStorageMode(StorageModePrivate);
    auto lookupTexture = NS::TransferPtr(_pDevice->newTexture(lookupTextureDescriptor));
    lookupTexture->setLabel(NS::String::string("DFG Lookup Table (GGX)", NS::UTF8StringEncoding));
    
    // 3. Pass 1: Equirect -> Cube
    // FIX: commandBuffer returns autoreleased -> No TransferPtr
    auto commandBuffer = _pCommandQueue->commandBuffer();
    // FIX: computeCommandEncoder returns autoreleased -> No TransferPtr
    auto computeCommandEncoder = commandBuffer->computeCommandEncoder();
    
    computeCommandEncoder->setComputePipelineState(_pEquirectToCubePSO.get());
    computeCommandEncoder->setTexture(equirectTexture.get(), 1);
    computeCommandEncoder->setTexture(sourceCubeTexture.get(), 0);
    computeCommandEncoder->dispatchThreads(MTL::Size::Make(sourceCubeSize, sourceCubeSize, 6),
                                           MTL::Size::Make(32, 32, 1));
    computeCommandEncoder->endEncoding();
    commandBuffer->commit();
    
    // 4. Mipmaps & All Other Computed Passes
    commandBuffer = _pCommandQueue->commandBuffer();
    
    auto mipmapCommandEncoder = commandBuffer->blitCommandEncoder();
    mipmapCommandEncoder->generateMipmaps(sourceCubeTexture.get());
    mipmapCommandEncoder->endEncoding();
    
    computeCommandEncoder = commandBuffer->computeCommandEncoder();
    
    struct PrefilteringParams {
        uint32_t distribution;
        uint32_t sampleCount;
        float roughness;
        float lodBias;
        float cubemapSize;
    };
    
    int mipLevelCount = 5;
    NS::UInteger levelSize = specularCubeSize;
    
    for (int mipLevel = 0; mipLevel < mipLevelCount; mipLevel++) {
        // newTextureView starts with 'new' -> Retained -> TransferPtr OK
        auto levelView = NS::TransferPtr(specularCubeTexture->newTextureView(workingPixelFormat, TextureTypeCube, NS::Range(mipLevel, 1), NS::Range(0, 6)));
        
        PrefilteringParams params;
        params.distribution = 1; // GGX
        params.sampleCount = (uint32_t)specularSampleCount;
        params.roughness = (float)mipLevel / (float)(mipLevelCount - 1);
        params.lodBias = 0.0f;
        params.cubemapSize = (float)sourceCubeSize;
        
        computeCommandEncoder->setComputePipelineState(_pPrefilteringPSO.get());
        computeCommandEncoder->setBytes(&params, sizeof(params), 0);
        computeCommandEncoder->setTexture(sourceCubeTexture.get(), 0);
        computeCommandEncoder->setTexture(levelView.get(), 1);
        computeCommandEncoder->dispatchThreads(MTL::Size::Make(levelSize, levelSize, 6),
                                               MTL::Size::Make(32, 32, 1));
        
        levelSize >>= 1;
    }
    
    // Diffuse Pass
    PrefilteringParams diffuseParams;
    diffuseParams.distribution = 0; // Lambert
    diffuseParams.sampleCount = (uint32_t)diffuseSampleCount;
    diffuseParams.roughness = 0.0f;
    diffuseParams.lodBias = 0.0f;
    diffuseParams.cubemapSize = (float)sourceCubeSize;
    
    computeCommandEncoder->setComputePipelineState(_pPrefilteringPSO.get());
    computeCommandEncoder->setBytes(&diffuseParams, sizeof(diffuseParams), 0);
    computeCommandEncoder->setTexture(sourceCubeTexture.get(), 0);
    computeCommandEncoder->setTexture(diffuseCubeTexture.get(), 1);
    computeCommandEncoder->dispatchThreads(MTL::Size::Make(diffuseCubeSize, diffuseCubeSize, 6),
                                           MTL::Size::Make(32, 32, 1));
    
    // LUT Pass
    uint32_t lutSamples = (uint32_t)lutSampleCount;
    computeCommandEncoder->setComputePipelineState(_pLookupTablePSO.get());
    computeCommandEncoder->setBytes(&lutSamples, sizeof(lutSamples), 0);
    computeCommandEncoder->setTexture(lookupTexture.get(), 0);
    computeCommandEncoder->dispatchThreads(MTL::Size::Make(lookupTableSize, lookupTableSize, 1),
                                           MTL::Size::Make(32, 32, 1));
    computeCommandEncoder->endEncoding();
    
    // Finalize
    auto readyEvent = NS::TransferPtr(_pDevice->newEvent());
    commandBuffer->encodeSignalEvent(readyEvent.get(), 1);
    commandBuffer->commit();
    
    return new ImageBasedLight(diffuseCubeTexture, specularCubeTexture, mipLevelCount, lookupTexture, readyEvent);
}
